feat: uninstall the test app once and install that again when MismatchedApplicationIdentifierEntitlement installation error occurs

KazuCocoa opened the pull request:
I saw an app installation got failed as mismatch provisioning profile identifier error even the OS did not have the given bundle id on the device via `isAppInstalled` check.

So, according to `isAppInstalled`, the app should not be on the device but an installation error occurred since the app was installed with another provisioning profile.

Not clear the reason, but based on the device log, potentially the OS internally cached some information. Some installation steps worked, but haven't been completed. The uninstall (device.remove) could help to clear such cache.


Another note in my investigation related to this, if the app under test remained in the device as an offload app (maybe not "cache"), the app is not "present" on the device (`isAppInstalled` is false, expected based on the available information) but the OS keeps the data. `driver.remove` uninstalls all of them for sure.

Thus, current fullReset/enforceAppInstall capabilities won't clear the local data if the application remains as offline application. Users must call `mobile: removeApp` explicitly to call the remove method for sure,

mykola-mokhnach commented on lib/real-device-management.js:
```diff
@@ -74,7 +74,9 @@ async function installToRealDevice(device, app, bundleId, opts) {
 
   const {skipUninstall, strategy, timeout} = opts ?? {};
 
-  if (!skipUninstall && bundleId && (await device.isAppInstalled(bundleId))) {
```
this looks like a workaround to me. Can the isAppInstalled API be fixed instead to return a proper value?

KazuCocoa replied:
Ideally, yes, I'd love to. I don't have the issue environment to try something, so if i can get such device, I'd love to.


One weird thing was this behavior occurred after uninstalling the app via `ideviceinstaller` for example.
https://github.com/appium/appium-xcuitest-driver/blob/ea9f383ee512cd70170f909f9a13ca2918957406/lib/ios-deploy.js#L181 this filter itself seems correct. The issue happened before https://github.com/appium/appium-xcuitest-driver/pull/1753 , thus probably `lookupApplications` by `usbmuxd` itself returned unexpected data potentially.


The device log also had `Pease try again later` like this error message for the installation failure message, thus potentially it needed more time to complete some internal process.

KazuCocoa replied:
It seems like we need this change.

The "offload" app https://github.com/appium/appium-xcuitest-driver/pull/2050#issuecomment-1740380759 seems not present in a couple of third-party libs as well, but the remove https://github.com/appium/appium-xcuitest-driver/blob/91ec8bf5b6bc68215cc9e05f09f09be3a03c074a/lib/ios-deploy.js#L28 was able to uninstall the app completely from the device even that case.

So, we should address this behavior in docs as well

KazuCocoa replied:
[9dc2288](https://github.com/appium/appium-xcuitest-driver/pull/2050/commits/9dc2288ed6ac0c3fc535827fc88ad4cf0e92d00a) to address this offload app behavior

KazuCocoa replied:
(From another thread to move to correct place)

`skipUninstall` is considered by `isAppInstalled` in `checkAutInstallationState`, so the `(await device.isAppInstalled(bundleId)` removal itself is maybe ok since `skipUninstall` includes the check. It will reduce ~0.2 sec extra time to check `isAppInstalled`

mykola-mokhnach commented on lib/real-device-management.js:
```diff
@@ -74,7 +74,9 @@ async function installToRealDevice(device, app, bundleId, opts) {
+  if (!skipUninstall && bundleId) {
     log.debug(`Reset requested. Removing app with id '${bundleId}' from the device`);
     await device.remove(bundleId);
   }
```
perhaps, we may still want to install, then analyze the error and retry

KazuCocoa replied:
Yea, still we observe the same error after calling uninstall command explicitly before install, the next would be retrial

mykola-mokhnach replied:
Sorry, I am not following. the sentence above mentions

>  `async remove`  was able to uninstall the app completely from the device even that case.

and then here

> still we observe the same error after calling uninstall command explicitly before install

I would like to keep the current logic if possible as reinstalling of an app is an expensive operation, which slows down session startup

KazuCocoa replied:
Oh, sorry. Please ignore "still we observe the same error after calling uninstall command explicitly before install" case since that was potentially not correct as ambiguous situation.

I'll update this pr title also for offload case to focus on this.

Btw, `skipUninstall` is considered by `isAppInstalled` in `checkAutInstallationState`, so the `(await device.isAppInstalled(bundleId)` removal itself is ok since `skipUninstall` includes the check. It will reduce ~0.2 sec extra time to check `isAppInstalled`

KazuCocoa commented:
I wondered if like this offloaded app was in a similar situation. The app was not listed in the result of `lookupApplications`, but existed in the device. I don't have an app to try to install with another profile etc, but the app existed in IPad Storage section in the settings app. So the OS itself was detected.
![Screenshot 2023-09-28 at 11 30 14 PM](https://github.com/appium/appium-xcuitest-driver/assets/5511591/939dab64-7439-4258-bdc1-a14e31ac3bec)

mykola-mokhnach commented on docs/execute-methods.md:
```diff
@@ -108,7 +108,7 @@ strategy | string | no | One of possible app installation strategies on real dev
 
 Checks whether the given application is installed on the device under test.
-Offload app could be handled as not installed.
+An offload application could be handled as not installed.
```
maybe it makes sence to make it to a link. Not everybody knows what `offload application` means

KazuCocoa replied:
Yes, I'll add link later

KazuCocoa commented:
Updated some changes, especially offload application case, that requires explicit `mobile: removeApp` to remove the local application data in current implementation basis

I'll update further later.

KazuCocoa commented:
On my local testing, I observed that when app installation failed, such a case also the  "offload" state happened.

So, the most easiest way to reproduce this is:

1. failed to install an app because invalid provisioning profile (from a fresh installation)
2. Then, the failed to install app remained on the device as "offload" app icon
    - Then, the app did not present via `usbmuxd` info
3. Install the same bundle id with another signature
4. The issue happens

So, actually this case requires app delete explicitly to avoid before the app installation. I'll update this branch further

KazuCocoa commented:
Updated.
So, the uninstall/install occurs only when `MismatchedApplicationIdentifierEntitlement` error occurs.
I also addressed the remaining data for offload state -> install succession case in enforeceInstall/fullReset in the documentation

mykola-mokhnach commented on docs/capabilities.md:
```diff
@@ -12,7 +12,7 @@ Capability&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbs
 `appium:udid` | UDID of the device to be tested. Could be retrieved from Xcode->Window->Devices and Simulators window. Always set this capability if you run parallel tests or use a real device to run your tests.
 `appium:noReset` | Prevents the device to be reset before the session startup if set to `true`. This means that the application under test is not going to be terminated neither its data cleaned. `false` by default
-`appium:fullReset` | Being set to `true` always enforces the application under test to be fully uninstalled before starting a new session. `false` by default
+`appium:fullReset` | Being set to `true` always enforces the application under test to be fully uninstalled before starting a new session. If the application under test remains on the device as an [offload application](https://discussions.apple.com/thread/254887240), the application data will remain. Please use [`mobile: removeApp`](execute-methods.md#mobile-removeapp) to ensure the application d ...
```
- not sure if this comment makes sence for simulators as we perform the reset for the whole device
- for real devices it would be better to just add a link to a troubleshooting article

mykola-mokhnach commented on docs/capabilities.md:
```diff
@@ -25,7 +25,7 @@ Capability | Description
 `appium:bundleId` | Bundle identifier of the app under test, for example `com.mycompany.myapp`. The capability value is calculated automatically if `app` is provided. If neither `app` or `bundleId` capability is provided then XCUITest driver starts from the Home screen.
 `appium:app` | Full path to the application to be tested (the app must be located on the same machine where the server is running). `.ipa` and `.app` application extensions are supported. Zipped `.app` bundles are supported as well. Could also be an URL to a remote location. If neither of the `app` or `bundleId` capabilities are provided then the driver starts from the Home screen and expects the ...
-`appium:enforceAppInstall` | If set to `false` it will make xcuitest driver to verify whether the app version currently installed on the device under test is older than the one, which is provided as `appium:app` value. No app reinstall is going to happen if the candidate app has the same or older version number than the already installed copy of it. The version number used for comparison must be  ...
+`appium:enforceAppInstall` | If set to `false` it will make xcuitest driver to verify whether the app version currently installed on the device under test is older than the one, which is provided as `appium:app` value. No app reinstall is going to happen if the candidate app has the same or older version number than the already installed copy of it. The version number used for comparison must be  ...
```
same here. Simply link a troubleshooting article

mykola-mokhnach commented on docs/execute-methods.md:
```diff
@@ -122,6 +123,13 @@ Either `true` or `false`
 Removes the given application from the device under test.
+An [offload application]((https://discussions.apple.com/thread/254887240)) also can be removed.
+
+Note that `appium:fullReset` or `appium:enforceAppInstall` capability may not uninstall such offload application state in a new session request.
```
same as above

mykola-mokhnach commented on lib/driver.js:
```diff
@@ -1610,7 +1610,9 @@ class XCUITestDriver extends BaseDriver {
     } else {
-      this.log.info(`App '${bundleId}' is not installed on the device yet`);
+      this.log.info(`App '${bundleId}' is not installed on the device yet, ` +
+      'or the application is offload state, which might keep the local data but the application ' +
```
...is not installed yet or it has an offload copy, which cannot be detected.

mykola-mokhnach commented on lib/driver.js:
```diff
@@ -1684,11 +1686,14 @@ class XCUITestDriver extends BaseDriver {
     const {install, skipUninstall} = await this.checkAutInstallationState();
     if (install) {
       if (this.isRealDevice()) {
+        const shouldEnforceUninstall = this.opts.enforceAppInstall !== false || this.opts.fullReset;
```
I am still not quite convinced why we need this flag. I thought we always perform uninstall if skipUninstall is not true

KazuCocoa replied:
That is to keep current behavior when  `enforceAppInstall: false` and `fullReset: false`, that expect not clear the device local state. So, if the device already had offload app for example, then the new session request may want to get the error without removing the app once and attempting to install it again.


`skipUninstall` can be true when wasAppInstalled is false like this offload app, or not installed yet. 
```
    if (enforceAppInstall !== false || fullReset || !wasAppInstalled) {
      return {
        install: true,
        skipUninstall: !wasAppInstalled,
      };
    }
```

Then, if the session does not require clearing the local (`enforceAppInstall: false` or `fullReset: false`), the install failure won't call `driver.remove` and `driver.install` even if the error is the `MismatchedApplicationIdentifierEntitlement`. Then, the current implementation raises an error

mykola-mokhnach replied:
> Then, if the session does not require clearing the local (enforceAppInstall: false or fullReset: false), the install failure won't call driver.remove and driver.install even if the error is the MismatchedApplicationIdentifierEntitlement

I am not quite sure why driver.remove won't be called in such case. We always try to remove an app if install fails with the particular error though

mykola-mokhnach commented on lib/real-device-management.js:
```diff
@@ -72,15 +73,71 @@ async function installToRealDevice(device, app, bundleId, opts) {
+      throw e;
+    }
+
+    if (!bundleId) {
```
I assume this won't happen. If the app path was given then we can always read the bundle identifier

KazuCocoa replied:
Yea, but the lint claimed the input could be undefined. Let me check typehint

KazuCocoa replied:
I have updated `unfedined` usage in the caller side in https://github.com/appium/appium-xcuitest-driver/pull/2050/commits/00ed99fdd75d54d74ab1971f5998ed2173b8b94d

KazuCocoa replied:
The changes is now in https://github.com/appium/appium-xcuitest-driver/pull/2054

mykola-mokhnach commented on lib/real-device-management.js:
```diff
@@ -72,15 +73,71 @@ async function installToRealDevice(device, app, bundleId, opts) {
+    // We cannot recover 'ApplicationVerificationFailed' situation since this reason is clearly the app's provisioning profile was invalid.
+    // [XCUITest] Error installing app '/path/to.app': Unexpected data: {"Error":"ApplicationVerificationFailed","ErrorDetail":-402620395,"ErrorDescription":"Failed to verify code signature of /path/to.app : 0xe8008015 (A valid provisioning profile for this executable was not found.)"}
+
+    if (shouldUninstallApp) {
```
I would say this is an overkill

lets just keep it simple

KazuCocoa replied:
Sure. I have changed the removal's loglevel to `info` to not lost the removal info that needs to keep the info instead of here

mykola-mokhnach commented on lib/real-device-management.js:
```diff
@@ -72,15 +73,71 @@ async function installToRealDevice(device, app, bundleId, opts) {
+    // by uninstalling the device's app bundle id explicitly regard less the app exists on the device or not (e.g. offload app).
+    // [XCUITest] Error installing app '/path/to.app': Unexpected data: {"Error":"MismatchedApplicationIdentifierEntitlement","ErrorDescription":"Upgrade's application-identifier entitlement string (TEAM_ID.com.kazucocoa.example) does not match installed application's application-identifier string (ANOTHER_TEAM_ID.com.kazucocoa.example); rejecting upgrade."}
+    if (!shouldUninstallApp && e.message.includes('MismatchedApplicationIdentifierEntitlement')) {
+      log.info(`MismatchedApplicationIdentifierEntitlement error occurred because '${bundleId}' might remain as ` +
```
How about:
```
The application identified by '{bundleId}' cannot be installed because it might be already cached on the device, probably with a different signature. Will try to remove it and install a new copy. Original error: ${e.message}'
```

mykola-mokhnach commented on docs/capabilities.md:
```diff
@@ -12,7 +12,7 @@ Capability&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbs
 `appium:udid` | UDID of the device to be tested. Could be retrieved from Xcode->Window->Devices and Simulators window. Always set this capability if you run parallel tests or use a real device to run your tests.
 `appium:noReset` | Prevents the device to be reset before the session startup if set to `true`. This means that the application under test is not going to be terminated neither its data cleaned. `false` by default
-`appium:fullReset` | Being set to `true` always enforces the application under test to be fully uninstalled before starting a new session. The application data would remain for real devices if the application under test remained on the device under a certain situation. Please check [troubleshooting](troubleshooting.md#clear-the-application-local-data-explicitly-for-real-devices) to clear the appl ...
+`appium:fullReset` | Being set to `true` always enforces the application under test to be fully uninstalled before starting a new session. The application data would remain for real devices if the application under test remained on the device under a certain situation. Please check [troubleshooting](troubleshooting.md#clear-the-application-local-data-explicitly-for-real-devices) for more details  ...
```
to clear the application local data for sure -> regarding obsolete application data cleanup on real devices

mykola-mokhnach replied:
same below

mykola-mokhnach commented on lib/real-device-management.js:
```diff
@@ -67,20 +68,51 @@ async function runRealDeviceReset(device, opts) {
+    await device.install(app, timeout, strategy);
+    log.debug('The app has been installed successfully.');
+  } catch (e) {
+    // Do nothing if the session did not require to enforce uninstalling the application
```
same as above. I would say it would be fine to always retry install by default and not have this condition
we could change this later if people complain about that, but altogether I try to always target positive scenarios where people actually want to have their app successfully installed rather than a test failure

KazuCocoa replied:
Okay, i just wanted to keep current behavior unless the request requires to enforce uninstallation explicitly. No other strong opinion I have. Then, I'll update title as well as feat

KazuCocoa replied:
removed in the latest code

mykola-mokhnach commented on docs/troubleshooting.md:
```diff
@@ -12,6 +12,21 @@ title: Troubleshooting
+
+## Clear the application local data explicitly for real devices
+
+iOS real device could have a situation that has application data locally but the application package is not on the device. It can occur an [offload application](https://discussions.apple.com/thread/254887240) state, cached application state or when an application installation fails. An example of the installation failure is `ApplicationVerificationFailed` by invalid provisioning profile.
```
There might be a situation where application data is present on the device although application itself is not installed. This could happen if:
- The app is in offload state
- The application state is cached
- There was an unexpected failure while installing the app. An example of such failure is the `ApplicationVerificationFailed` which happens while installing an app signed with an invalid provisioning profile.

Under the circumstances above the application identifier won't be listed in the `mobile: listApps` output neither deleted by `mobile: isAppInstalled` API. Setting `fullReset` or `enforceAppInstall` capabilities to `true` won't help to clear this data too.

The only way to completely get rid of the cached app data is to call the `mobile: removeApp` API with the appropriate bundle identifier.

The driver automatically tries to resolve application installs that failed because of the obsolete data presence, although, if you explicitly ask it to not perform the app uninstall then consider calling `mobile: removeApp` beforehand.

KazuCocoa commented on lib/real-device-management.js:
```diff
@@ -67,20 +68,46 @@ async function runRealDeviceReset(device, opts) {
 async function installToRealDevice(device, app, bundleId, opts) {
-  if (!device.udid || !app) {
-    log.debug('No device id or app, not installing to real device.');
+  if (!device.udid || !app || !bundleId) {
```
https://github.com/appium/appium-xcuitest-driver/pull/2054 must merge first since this check will make the current installation fail

mykola-mokhnach replied:
I have merged the above PR. Please rebase

mykola-mokhnach commented on docs/capabilities.md:
```diff
@@ -12,7 +12,7 @@ Capability&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbs
 `appium:udid` | UDID of the device to be tested. Could be retrieved from Xcode->Window->Devices and Simulators window. Always set this capability if you run parallel tests or use a real device to run your tests.
 `appium:noReset` | Prevents the device to be reset before the session startup if set to `true`. This means that the application under test is not going to be terminated neither its data cleaned. `false` by default
-`appium:fullReset` | Being set to `true` always enforces the application under test to be fully uninstalled before starting a new session. `false` by default
+`appium:fullReset` | Being set to `true` always enforces the application under test to be fully uninstalled before starting a new session. The application data would remain for real devices if the application under test remained on the device under a certain situation. Please check [troubleshooting](troubleshooting.md#clear-the-application-local-data-explicitly-for-real-devices) for more details  ...
```
The application data would remain for real devices if the application under test remained on the device under a certain situation. > The application data might be cached on real devices under particular circumstances.

mykola-mokhnach commented on docs/execute-methods.md:
```diff
@@ -123,6 +124,9 @@ Either `true` or `false`
 Removes the given application from the device under test.
 
+An [offload application]((https://discussions.apple.com/thread/254887240)) also can be removed.
+Please check [Clear the application local data explicitly for real devices](troubleshooting.md#clear-the-application-local-data-explicitly-for-real-devices) to ensure how to clear the application local data.
```
to ensure app data clenup.

mykola-mokhnach commented on docs/troubleshooting.md:
```diff
@@ -12,6 +12,26 @@ title: Troubleshooting
+The driver automatically tries to resolve application installs that failed because of `MismatchedApplicationIdentifierEntitlement`, although, if you explicitly ask it to not perform the application uninstall then consider calling [`mobile: removeApp`](execute-methods.md#mobile-removeapp) beforehand `MismatchedApplicationIdentifierEntitlement` error occurs only when the previously installed applic ...
+
+An example to start a session with explicit application uninstallation to clear the local data.
+1. Start a session without `appium:app` nor `appium:bundleId`
```
nor -> and

mykola-mokhnach commented on lib/commands/app-management.js:
```diff
@@ -52,6 +53,7 @@ export default {
   },
   /**
    * Removes/uninstalls the given application from the device under test.
+   * Offload app also can be removed.
```
Offload app data could also be removed

mykola-mokhnach commented on lib/driver.js:
```diff
@@ -1610,7 +1610,8 @@ class XCUITestDriver extends BaseDriver {
       }
     } else {
-      this.log.info(`App '${bundleId}' is not installed on the device yet`);
+      this.log.info(`App '${bundleId}' is not installed yet or it has an offload, ` +
```
and cannot be detected

mykola-mokhnach commented on lib/real-device-management.js:
```diff
@@ -67,20 +68,46 @@ async function runRealDeviceReset(device, opts) {
+    // by uninstalling the device's app bundle id explicitly regard less the app exists on the device or not (e.g. offload app).
+    // [XCUITest] Error installing app '/path/to.app': Unexpected data: {"Error":"MismatchedApplicationIdentifierEntitlement","ErrorDescription":"Upgrade's application-identifier entitlement string (TEAM_ID.com.kazucocoa.example) does not match installed application's application-identifier string (ANOTHER_TEAM_ID.com.kazucocoa.example); rejecting upgrade."}
+    if (!shouldUninstallApp && e.message.includes('MismatchedApplicationIdentifierEntitlement')) {
+      log.info(`The application identified by '{bundleId}' cannot be installed because it might ` +
```
missing dollar sign before `{`

KazuCocoa commented:
Updated

mykola-mokhnach commented on lib/driver.js:
```diff
@@ -1610,7 +1610,8 @@ class XCUITestDriver extends BaseDriver {
     } else {
-      this.log.info(`App '${bundleId}' is not installed on the device yet`);
+      this.log.info(`App '${bundleId}' is not installed yet or it has an offload and ` +
+        'cannot be detected, which might keep the local data, cannot be detected.');
```
`cannot be detected` is repeated twice

mykola-mokhnach commented on lib/real-device-management.js:
```diff
@@ -67,20 +68,46 @@ async function runRealDeviceReset(device, opts) {
   }
 
   const {skipUninstall, strategy, timeout} = opts ?? {};
+  const shouldUninstallApp = !skipUninstall && bundleId;
```
bundleId is always truthy because of the check above

mykola-mokhnach commented on lib/real-device-management.js:
```diff
@@ -67,20 +68,46 @@ async function runRealDeviceReset(device, opts) {
+    // [XCUITest] Error installing app '/path/to.app': Unexpected data: {"Error":"MismatchedApplicationIdentifierEntitlement","ErrorDescription":"Upgrade's application-identifier entitlement string (TEAM_ID.com.kazucocoa.example) does not match installed application's application-identifier string (ANOTHER_TEAM_ID.com.kazucocoa.example); rejecting upgrade."}
+    if (!shouldUninstallApp && e.message.includes('MismatchedApplicationIdentifierEntitlement')) {
+      log.info(`The application identified by '${bundleId}' cannot be installed because it might ` +
+        `be already cached on the device, probably with a different signature by some reasons. ` +
```
probably with a different signature by some reasons. > probably with a different signature.

mykola-mokhnach commented on test/unit/real-device-management-specs.js:
```diff
@@ -0,0 +1,122 @@
+import chai from 'chai';
```
👍

mykola-mokhnach commented on docs/troubleshooting.md:
```diff
@@ -12,6 +12,26 @@ title: Troubleshooting
+
+The driver automatically tries to resolve application installs that failed because of `MismatchedApplicationIdentifierEntitlement`, although, if you explicitly ask it to not perform the application uninstall then consider calling [`mobile: removeApp`](execute-methods.md#mobile-removeapp) beforehand `MismatchedApplicationIdentifierEntitlement` error occurs only when the previously installed applic ...
+
+An example to start a session with explicit application uninstallation to clear the local data.
```
Below are example steps of the manual approach mentioned above:

mykola-mokhnach commented on lib/real-device-management.js:
```diff
@@ -67,20 +68,45 @@ async function runRealDeviceReset(device, opts) {
+
+    // We cannot recover 'ApplicationVerificationFailed' situation since this reason is clearly the app's provisioning profile was invalid.
+    // [XCUITest] Error installing app '/path/to.app': Unexpected data: {"Error":"ApplicationVerificationFailed","ErrorDetail":-402620395,"ErrorDescription":"Failed to verify code signature of /path/to.app : 0xe8008015 (A valid provisioning profile for this executable was not found.)"}
+    throw e;
```
I would probably put this part into a guard condition instead as it results into a shorter block:

```
if (!skipUninstall || !e.message.includes('MismatchedApplicationIdentifierEntitlement') {
  throw e;
}
....
```
